### PR TITLE
[SPARK-35452][PYTHON] Introduce ArrayOps, MapOps and StructOps

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?

The PR is proposed to introduce ArrayOps, MapOps and StructOps to handle data-type-based operations for StructType, ArrayType, and MapType separately.

### Why are the changes needed?

StructType, ArrayType, and MapType are not accepted by DataTypeOps now.

We should handle these complex types. Among them:

- ArrayType supports concatenation: for example, ps.Series([[1,2,3]]) + ps.Series([[4,5,6]]) should work the same as pd.Series([[1,2,3]]) + pd.Series([[4,5,6]]), as concatenation.

- StructOps will be helpful to make to/from pandas conversion data-type-based.

### Does this PR introduce _any_ user-facing change?

Yes.

Before the change:
```py
>>> import pyspark.pandas as ps
>>> from pyspark.pandas.config import set_option
>>> set_option("compute.ops_on_diff_frames", True)
>>> ps.Series([[1, 2, 3]]) + ps.Series([[0.4, 0.5]])
Traceback (most recent call last):
...
TypeError: Type object was not understood.
>>> ps.Series([[1, 2, 3]]) + ps.Series([[4, 5]])
Traceback (most recent call last):
...
TypeError: Type object was not understood.
>>> ps.Series([[1, 2, 3]]) + ps.Series([['x']])
Traceback (most recent call last):
...
TypeError: Type object was not understood.
```

After the change:
```py
>>> import pyspark.pandas as ps
>>> from pyspark.pandas.config import set_option
>>> set_option("compute.ops_on_diff_frames", True)
>>> ps.Series([[1, 2, 3]]) + ps.Series([[0.4, 0.5]])
0    [1.0, 2.0, 3.0, 0.4, 0.5]                                                  
dtype: object
>>> ps.Series([[1, 2, 3]]) + ps.Series([[4, 5]])
0    [1, 2, 3, 4, 5]
dtype: object
>>> ps.Series([[1, 2, 3]]) + ps.Series([['x']])
Traceback (most recent call last):
...
TypeError: Concatenation can only be applied to arrays of the same type
```

### How was this patch tested?

Unit tests.

Keyword: SPARK-35337

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

File: python/pyspark/pandas/data_type_ops/complex_ops.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+from pyspark.pandas.data_type_ops.base import DataTypeOps
+from pyspark.sql import functions as F
+from pyspark.sql.types import ArrayType, NumericType
+
+if TYPE_CHECKING:
+    from pyspark.pandas.indexes import Index  # noqa: F401 (SPARK-34943)
+    from pyspark.pandas.series import Series  # noqa: F401 (SPARK-34943)
+
+
+class ComplexOps(DataTypeOps):
```
Comment (ueshin) [inline python/pyspark/pandas/data_type_ops/complex_ops.py:30]: Shall we have `MapOps` and `StructOps` separately instead of having a common `ComplexOps`?

Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/complex_ops.py]: Makes sense! `ComplexOps` is removed. Would you prefer ArrayOps, MapOps and StructOps to be in separate files?

File: python/pyspark/pandas/data_type_ops/complex_ops.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class MapOps(DataTypeOps):
+    """
+    The class for binary operations of pandas-on-Spark objects with MapType.
+    """
+
+    @property
+    def pretty_name(self) -> str:
+        return 'PySpark MapType'
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/data_type_ops/complex_ops.py:64]: Does the name look good?

File: python/pyspark/pandas/data_type_ops/complex_ops.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/pandas/data_type_ops/complex_ops.py]: I think it's fine to have them in a single file for now.

File: python/pyspark/pandas/data_type_ops/complex_ops.py (hunk 2 above)
Comment (ueshin) [reply python/pyspark/pandas/data_type_ops/complex_ops.py:64]: `maps` and `structs` should be fine for now?

Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/complex_ops.py]: Sounds good! Renamed.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**